source/loader/layers/sanitizer/ur_sanddi.cpp

```diff
@@ -316,6 +316,841 @@ __urdlllocal ur_result_t UR_APICALL urContextRelease(
+        return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    }
+
+    context.logger.debug("==== urEnqueueMemUnmap");
+
+    if (auto MemBuffer = context.interceptor->getMemBuffer(hMem)) {
+        MemBuffer::Mapping Mapping{};
+        {
+            std::scoped_lock<ur_shared_mutex> Guard(MemBuffer->Mutex);
+            auto It = MemBuffer->Mappings.find(pMappedPtr);
+            UR_ASSERT(It != MemBuffer->Mappings.end(),
+                      UR_RESULT_ERROR_INVALID_VALUE);
+            Mapping = It->second;
+            MemBuffer->Mappings.erase(It);
+        }
+
+        // Write back mapping memory data to device and release mapping memory
+        // if we allocated a host USM. But for now, UR doesn't support event
+        // call back, we can only do blocking copy here.
+        char *DstHandle = nullptr;
+        ur_context_handle_t Context = GetContext(hQueue);
+        ur_device_handle_t Device = GetDevice(hQueue);
+        UR_CALL(MemBuffer->getHandle(Device, DstHandle));
+        UR_CALL(context.urDdiTable.Enqueue.pfnUSMMemcpy(
+            hQueue, true, DstHandle + Mapping.Offset, pMappedPtr, Mapping.Size,
+            numEventsInWaitList, phEventWaitList, phEvent));
+
+        if (!MemBuffer->HostPtr) {
+            UR_CALL(context.interceptor->releaseMemory(Context, pMappedPtr));
+        }
+    } else {
+        UR_CALL(pfnMemUnmap(hQueue, hMem, pMappedPtr, numEventsInWaitList,
+                            phEventWaitList, phEvent));
+    }
+
+    return UR_RESULT_SUCCESS;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Intercept function for urKernelCreate
+__urdlllocal ur_result_t UR_APICALL urKernelCreate(
+    ur_program_handle_t hProgram, ///< [in] handle of the program instance
+    const char *pKernelName,      ///< [in] pointer to null-terminated string.
+    ur_kernel_handle_t
+        *phKernel ///< [out] pointer to handle of kernel object created.
+) {
+    auto pfnCreate = context.urDdiTable.Kernel.pfnCreate;
+
+    if (nullptr == pfnCreate) {
+        return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    }
+
+    context.logger.debug("==== urKernelCreate");
+
+    UR_CALL(pfnCreate(hProgram, pKernelName, phKernel));
+    UR_CALL(context.interceptor->insertKernel(*phKernel));
+
+    return UR_RESULT_SUCCESS;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Intercept function for urKernelRelease
+__urdlllocal ur_result_t urKernelRelease(
+    ur_kernel_handle_t hKernel ///< [in] handle for the Kernel to release
+) {
+    auto pfnRelease = context.urDdiTable.Kernel.pfnRelease;
+
+    if (nullptr == pfnRelease) {
+        return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    }
+
+    context.logger.debug("==== urKernelRelease");
+    UR_CALL(pfnRelease(hKernel));
+
+    if (auto KernelInfo = context.interceptor->getKernelInfo(hKernel)) {
+        uint32_t RefCount;
+        UR_CALL(context.urDdiTable.Kernel.pfnGetInfo(
```

Comment:
This is not really thread-safe. If multiple `urKernelRelease` are done in parallel, multiple threads may read RefCount == 1. I think you might need to have your own RefCount like for MemBuffer.

Reply:
Thanks, you are right. Updated.